File: apollo_dump/_archived_20260605/scan_ic_diff.py
"""
scan_ic_diff.py — 练习场前后对比扫描 ItemCode
流程: scan1 → 进练习场 → scan2 → diff
"""
import sys, os, time, subprocess
from datetime import datetime
sys.stdout.reconfigure(encoding='utf-8')

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
LOG_FILE = os.path.join(SCRIPT_DIR, f'scan_diff_{datetime.now().strftime("%Y%m%d_%H%M%S")}.txt')
log_lines = []

def log(msg):
    line = f'[{time.strftime("%H:%M:%S")}] {msg}'
    print(line)
    log_lines.append(line)

def save_log():
    with open(LOG_FILE, 'w', encoding='utf-8') as f:
        f.write('\n'.join(log_lines))
    print(f'\n日志已保存: {LOG_FILE}')

def find_pid():
    try:
        r = subprocess.run(
            ['tasklist', '/fi', 'IMAGENAME eq FreeStyle.exe', '/fo', 'csv', '/nh'],
            capture_output=True, text=True, timeout=10
        )
        for line in r.stdout.strip().split('\n'):
            if 'FreeStyle.exe' in line:
                return int(line.split(',')[1].strip('"'))
    except Exception as e:
        log(f'查找进程失败: {e}')
    return None

def main():
    pid = find_pid()
    if not pid:
        log('FreeStyle.exe 未运行')
        return 1
    log(f'PID={pid}')

    import frida
    session = frida.attach(pid)
    log('Frida 已附加')

    js_path = os.path.join(SCRIPT_DIR, 'scan_ic_diff.js')
    with open(js_path, 'r', encoding='utf-8') as f:
        js_code = f.read()

    def on_msg(msg, data):
        if msg['type'] == 'error':
            log(f'[JS错误] {msg.get("description","")} @ line {msg.get("lineNumber","?")}')
            return
        if msg['type'] != 'send':
            return
        p = msg['payload']
        t = p.get('t', '')

        if t == 'ready':
            log(f'  {p["msg"]}')
        elif t == 'scan':
            log(f'  [{p["label"]}] 命中 {p["count"]} 处')
        elif t == 'diff':
            log(f'  对比: 只第一轮={p["only1"]} 只练习场={p["only2"]} 共有={p["both"]}')
        elif t == 'new':
            log(f'  [练习场新增 {p["n"]}] {p["addr"]}  当前值=0x{p["val"]:x} ({p["val"]})')
            ctx = p.get('ctx', '')
            if ctx:
                log(f'       上下文: {ctx}')
        elif t == 'changed':
            log(f'  [值变更 {p["n"]}] {p["addr"]}  {p["oldVal"]}→{p["newVal"]} (0x{p["newVal"]:x})')

    script = session.create_script(js_code)
    script.on('message', on_msg)
    script.load()

    log('')
    log('操作流程:')
    log('  1. 输入 scan1  → 第一轮扫描(大厅状态)')
    log('  2. 进练习场')
    log('  3. 输入 scan2  → 第二轮扫描(练习场状态)')
    log('  4. 输入 diff   → 对比差异')
    log('')

    try:
        while True:
            cmd = input('> ').strip().lower()
            if cmd in ('quit', 'q', 'exit'):
                break
            elif cmd == 'scan1':
                n = script.exports_sync.scan1()
                log(f'第一轮: {n} 处')
            elif cmd == 'scan2':
                n = script.exports_sync.scan2()
                log(f'第二轮: {n} 处')
            elif cmd == 'diff':
                result = script.exports_sync.diff()
                log(f'结果: {result}')
            elif cmd.startswith('read '):
                addr = cmd.split()[1]
                ctx = script.exports_sync.read(addr)
                log(f'{addr}: {ctx}')
            else:
                log('命令: scan1 / scan2 / diff / read <addr> / quit')
    except (KeyboardInterrupt, EOFError):
        pass

    script.unload()
    session.detach()
    log('已断开')
    save_log()
    return 0

if __name__ == '__main__':
    try:
        sys.exit(main())
    except Exception as e:
        log(f'异常: {e}')
        save_log()

File: apollo_dump/_archived_20260605/scan_ic_diff.js
// scan_ic_diff.js
// 分两轮扫描: 第一轮(当前状态) → 等用户操作 → 第二轮 → 对比差异
// 用法: 脚本加载后先扫一次, 用户进练习场, 输入 'scan' 再扫一次, 输入 'diff' 看差异

var SRC_IC = 50125461;
var icHex = ('0' + (SRC_IC & 0xFF).toString(16)).slice(-2) + ' ' +
            ('0' + ((SRC_IC >> 8) & 0xFF).toString(16)).slice(-2) + ' ' +
            ('0' + ((SRC_IC >> 16) & 0xFF).toString(16)).slice(-2) + ' ' +
            ('0' + ((SRC_IC >> 24) & 0xFF).toString(16)).slice(-2);

var scan1 = null;  // 第一轮结果
var scan2 = null;  // 第二轮结果

function doScan(label) {
    var gameMod = Process.getModuleByName('FreeStyle.exe');
    var hits = [];

    // 扫游戏模块
    Memory.scanSync(gameMod.base, gameMod.size, icHex).forEach(function(m) {
        hits.push({addr: m.address.toString(), inModule: true});
    });

    // 扫堆
    Process.enumerateRanges('rw-').forEach(function(range) {
        var rm = Process.findModuleByAddress(range.base);
        if (rm) return;
        try {
            Memory.scanSync(range.base, range.size, icHex).forEach(function(m) {
                hits.push({addr: m.address.toString(), inModule: false});
            });
        } catch(e) {}
    });

    send({t:'scan', label: label, count: hits.length});
    return hits;
}

// 暴露给 Python 调用的接口
rpc.exports = {
    scan1: function() {
        scan1 = doScan('第一轮(当前)');
        return scan1.length;
    },
    scan2: function() {
        scan2 = doScan('第二轮(练习场)');
        return scan2.length;
    },
    diff: function() {
        if (!scan1 || !scan2) return '请先执行两轮扫描';

        var set1 = {};
        scan1.forEach(function(h) { set1[h.addr] = true; });
        var set2 = {};
        scan2.forEach(function(h) { set2[h.addr] = true; });

        var only1 = [];  // 只在第一轮
        var only2 = [];  // 只在第二轮(练习场新增)
        var both = [];   // 两轮都有

        scan1.forEach(function(h) {
            if (!set2[h.addr]) only1.push(h.addr);
            else both.push(h.addr);
        });
        scan2.forEach(function(h) {
            if (!set1[h.addr]) only2.push(h.addr);
        });

        send({t:'diff', only1: only1.length, only2: only2.length, both: both.length});

        // 练习场新增的地址最关键 — 直接在 JS 内读上下文，不走 RPC
        only2.forEach(function(addr, i) {
            var ptr = new NativePointer(addr);
            var val = ptr.readU32();
            var ctx = '';
            try {
                var bytes = new Uint8Array(ptr.sub(8).readByteArray(24));
                for (var j = 0; j < bytes.length; j++) {
                    if (j === 8) ctx += '| ';
                    ctx += ('0' + bytes[j].toString(16)).slice(-2) + ' ';
                }
            } catch(e) { ctx = 'read error'; }
            send({t:'new', n:i+1, addr:addr, val:val, ctx:ctx});
        });

        // 检查两轮都有的地址，值是否变了
        both.forEach(function(addr, i) {
            var ptr = new NativePointer(addr);
            var val = ptr.readU32();
            if (val !== SRC_IC) {
                send({t:'changed', n:i+1, addr:addr, oldVal:SRC_IC, newVal:val});
            }
        });

        return {only1: only1.length, only2: only2.length, both: both.length};
    },
    read: function(addrStr) {
        var ptr = new NativePointer(addrStr);
        var ctx = new Uint8Array(ptr.sub(16).readByteArray(48));
        var hex = '';
        for (var i = 0; i < ctx.length; i++) {
            if (i === 16) hex += '| ';
            hex += ('0' + ctx[i].toString(16)).slice(-2) + ' ';
        }
        return hex;
    }
};

send({t:'ready', msg:'就绪。命令: scan1(第一轮) → 进练习场 → scan2(第二轮) → diff(对比)'});
